File: app/src/main/java/com/khmelenko/lab/travisclient/fragment/JobsFragment.java
package com.khmelenko.lab.travisclient.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.khmelenko.lab.travisclient.R;
import com.khmelenko.lab.travisclient.adapter.BranchesListAdapter;
import com.khmelenko.lab.travisclient.adapter.JobsListAdapter;
import com.khmelenko.lab.travisclient.network.response.Job;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Fragment with build jobs
 *
 * @author Dmytro Khmelenko
 */
public class JobsFragment extends Fragment {

    @Bind(R.id.jobs_recycler_view)
    RecyclerView mJobsRecyclerView;

    private JobsListAdapter mJobsListAdapter;
    private List<Job> mJobs = new ArrayList<>();

    private JobsListener mListener;

    /**
     * Creates new instance of the fragment
     *
     * @return Fragment instance
     */
    public static JobsFragment newInstance() {
        JobsFragment fragment = new JobsFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    public JobsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_jobs, container, false);
        ButterKnife.bind(this, view);

        mJobsRecyclerView.setHasFixedSize(true);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        mJobsRecyclerView.setLayoutManager(layoutManager);

        mJobsListAdapter = new JobsListAdapter(getContext(), mJobs);
        mJobsRecyclerView.setAdapter(mJobsListAdapter);

        return view;
    }


    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            mListener = (JobsListener) activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement JobsListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    /**
     * Sets the list of jobs
     *
     * @param jobs Jobs
     */
    public void setJobs(List<Job> jobs) {
        mJobs.clear();
        mJobs.addAll(jobs);

        mJobsListAdapter.notifyDataSetChanged();
    }

    /**
     * Interface for communication with this fragment
     */
    public interface JobsListener {

        /**
         * Handles job selection
         *
         * @param job Selected job
         */
        void onJobSelected(String job);
    }

}

File: app-v3/src/main/java/com/khmelenko/lab/varis/builddetails/JobsFragment.kt
package com.khmelenko.lab.varis.builddetails

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.khmelenko.lab.varis.R
import com.khmelenko.lab.varis.network.response.Job
import kotlinx.android.synthetic.main.fragment_jobs.jobs_recycler_view

/**
 * Fragment with build jobs
 *
 * @author Dmytro Khmelenko
 */
class JobsFragment : Fragment() {

    private lateinit var jobsListAdapter: JobsListAdapter
    private val jobs = ArrayList<Job>()

    private var listener: JobsListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_jobs, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        jobs_recycler_view.isNestedScrollingEnabled = false
        jobs_recycler_view.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(context)
        jobs_recycler_view.layoutManager = layoutManager

        jobs.clear()
        jobs.addAll(readJobsFromArgs())

        jobsListAdapter = JobsListAdapter(context!!, jobs) { position ->
            if (!jobs.isEmpty()) {
                val job = jobs[position]
                listener!!.onJobSelected(job)
            }
        }
        jobs_recycler_view.adapter = jobsListAdapter

        val metrics = resources.displayMetrics
        val itemHeight = (jobsListAdapter.itemHeight * metrics.density + 0.5).toInt()
        jobs_recycler_view.layoutParams.height = itemHeight * jobsListAdapter.itemCount
    }

    override fun onAttach(activity: Context) {
        super.onAttach(activity)
        try {
            listener = activity as JobsListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("$activity must implement JobsListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    /**
     * Sets the list of jobs
     *
     * @param jobs Jobs
     */
    fun setJobs(jobs: List<Job>) {
        this.jobs.clear()
        this.jobs.addAll(jobs)

        jobsListAdapter.notifyDataSetChanged()
    }

    private fun readJobsFromArgs(): List<Job> {
        val jobsAsString = arguments!!.getString(JOBS_KEY)
        val gson = Gson()
        val listType = object : TypeToken<List<Job>>() {}.type
        return gson.fromJson<List<Job>>(jobsAsString, listType)
    }

    /**
     * Interface for communication with this fragment
     */
    interface JobsListener {

        /**
         * Handles job selection
         *
         * @param job Selected job
         */
        fun onJobSelected(job: Job)
    }

    companion object {

        const val JOBS_KEY = "JOBS_KEY"

        /**
         * Creates new instance of the fragment
         *
         * @return Fragment instance
         */
        fun newInstance(jobs: List<Job>): JobsFragment {
            val fragment = JobsFragment()
            val args = Bundle()
            args.putString(JOBS_KEY, jobsToJson(jobs))
            fragment.arguments = args
            return fragment
        }

        private fun jobsToJson(jobs: List<Job>): String {
            val gson = Gson()
            return gson.toJson(jobs)
        }
    }

}
